src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -101,113 +108,102 @@ static ClioConfigDefinition ClioConfig = ClioConfigDefinition{
+{
+    for (auto const& p : pair) {
+        if (p.first.contains("[]"))
+            ASSERT(std::holds_alternative<Array>(p.second), "Value must be array if key has \"[]\"");
+        map_.insert(p);
+    }
+}
+
 ObjectView
 ClioConfigDefinition::getObject(std::string_view prefix, std::optional<std::size_t> idx) const
 {
-    std::string prefixWithDot = std::string(prefix) + ".";
+    auto const prefixWithDot = std::string(prefix) + ".";
     for (auto const& [mapKey, mapVal] : map_) {
-        if (mapKey.starts_with(prefixWithDot))
-            ASSERT(!mapKey.ends_with(".[]"), "Trying to retrieve an object when value is an Array");
-
-        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<ConfigValue>(mapVal))
-            return ObjectView{prefix, *this};
-
-        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<Array>(mapVal)) {
-            ASSERT(std::get<Array>(mapVal).size() > idx, "index provided is out of scope");
+        auto const hasPrefix = mapKey.starts_with(prefixWithDot);
+        if (idx.has_value() && hasPrefix && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx.value(), "Index provided is out of scope");
+            // we want to support getObject("array") and getObject("array.[]"), so we check if "[]" exists
+            if (!prefix.contains("[]"))
+                return ObjectView{prefixWithDot + "[]", idx.value(), *this};
             return ObjectView{prefix, idx.value(), *this};
         }
+        if (hasPrefix && !idx.has_value() && !mapKey.contains(prefixWithDot + "[]"))
+            return ObjectView{prefix, *this};
     }
-    throw std::invalid_argument("Key is not found in config");
+    ASSERT(false, "Key {} is not found in config", prefixWithDot);
+    return ObjectView{"", *this};
 }
 
 ArrayView
 ClioConfigDefinition::getArray(std::string_view prefix) const
 {
-    std::string key = std::string(prefix);
-    if (!prefix.contains(".[]"))
-        key += ".[]";
+    auto key = addBracketsForArrayKey(prefix);
 
     for (auto const& [mapKey, mapVal] : map_) {
-        if (mapKey.starts_with(key))
+        if (mapKey.starts_with(key)) {
+            ASSERT(
+                std::holds_alternative<Array>(mapVal), "Trying to retrieve Object or ConfigValue, instead of an Array "
+            );
             return ArrayView{key, *this};
-        ASSERT(!mapKey.starts_with(prefix), "Trying to retrieve an Array when value is an object");
+        }
     }
-    throw std::invalid_argument("Key is not found in config");
+    ASSERT(false, "Key {} is not found in config", key);
+    return ArrayView{"", *this};
+}
+
+bool
+ClioConfigDefinition::contains(std::string_view key) const
+{
+    return map_.contains(key);
+}
+
+bool
+ClioConfigDefinition::hasItemsWithPrefix(std::string_view key) const
+{
+    auto it = std::ranges::find_if(map_, [&key](auto const& pair) { return pair.first.starts_with(key); });
```

Comment:
Do we really need a mutable (or any) temporary here? 

It also could be possible to use `std::ranges::any_of` (https://compiler-explorer.com/z/W3hTjMPzM).
Unfortunately for some reason `std::ranges::contains` doesn't provide an overload that takes a user predicate 🤦 Would be the most suitable algorithm otherwise imo.